src/memkind_memtier.c

```diff
@@ -64,11 +74,38 @@ struct memtier_memory {
     struct memtier_tier_cfg *cfg; // Memory Tier configuration
 };
 
-static MEMKIND_ATOMIC size_t kind_alloc_size[MEMKIND_MAX_KIND];
+#define THREAD_BUCKETS (256U)
+
+static MEMKIND_ATOMIC size_t kind_alloc_size[MEMKIND_MAX_KIND][THREAD_BUCKETS];
+
+void memtier_reset_size(unsigned kind_id)
+{
+    unsigned b;
+    for (b = 0; b < THREAD_BUCKETS; ++b) {
+        memkind_atomic_set(kind_alloc_size[kind_id][b], 0);
+    }
+}
+
+// SplitMix64 hash
+static inline unsigned t_hash_64(void)
+{
+    uint64_t x = (uint64_t)pthread_self();
+    x += 0x9e3779b97f4a7c15;
+    x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9;
+    x = (x ^ (x >> 27)) * 0x94d049bb133111eb;
+    return (x ^ (x >> 31)) & (THREAD_BUCKETS - 1);
+}
+
+static inline void increment_alloc_size(unsigned kind_id, size_t size)
+{
+    unsigned b_id = t_hash_64();
```

Comment:
bucket_id instead of b_id would be a much more meaningful name

Reply:
Done.